### PR TITLE
fix: pick up upstream enhancements for tarfile reproducability (fixes #421)

### LINKED ISSUE #421: pick up upstream archive.py enhancements
AFAICT we are missing at least two enhancements to archive.py that improve determinism:
- https://bugzilla.mozilla.org/show_bug.cgi?id=1347582
- https://phabricator.services.mozilla.com/D165647

I ran into this while trying to reproduce an optimized taskgraph earlier today, and kept ending up with different tarfiles for the docker context, despite the files inside the archive being the same.

### PR DESCRIPTION
**Pull request by bhearsum**
This is mostly copy/pasting from https://searchfox.org/mozilla-central/source/python/mozbuild/mozpack/archive.py, with a couple of tweaks because we don't have the mozpack `file` classes here.
    
The enchancements come from https://bugzilla.mozilla.org/show_bug.cgi?id=1807872 and https://bugzilla.mozilla.org/show_bug.cgi?id=1347582.

### CODE REVIEW

File: src/taskgraph/util/archive.py [hunk 1]
```diff
@@ -12,6 +12,40 @@
...
+# What we do is subblass TarInfo so that if used on python >= 3.9, it reproduces the
+# behavior from python < 3.9.
+# Here's how it goes:
+# - the behavior in python >= 3.9 is the same as python < 3.9 when the type encoded
+# in the tarinfo is CHRTYPE or BLKTYPE.
+# - the value of the type is only compared in the context of choosing which behavior
+# to take
+# - we replace the type with the same value (so that using the value has no changes)
+# but that pretends to be the same as CHRTYPE so that the condition that enables the
+# old behavior is taken.
```
Comment (hneiva) [inline src/taskgraph/util/archive.py:32]: praise: Nice explanation!

File: src/taskgraph/util/archive.py [hunk 2]
```diff
@@ -12,6 +12,40 @@
...
+        return self == other
+
+
+class TarInfo(tarfile.TarInfo):
+    @staticmethod
+    def _create_header(info, format, encoding, errors):
+        info["type"] = HackedType(info["type"])
+        # ignore type checking because it looks like pyright complains because we're calling a
+        # non-public method
+        return tarfile.TarInfo._create_header(info, format, encoding, errors)  # type: ignore
```
Comment (hneiva) [inline src/taskgraph/util/archive.py:46]: nit: would it still complain if we used `self._create_header`?

**Review by hneiva — APPROVED** [review]
Minor nit, but LGTM
Nice hacking!

File: src/taskgraph/util/archive.py (hunk 2 above)
Comment (bhearsum) [reply src/taskgraph/util/archive.py:46]: AIUI, that would end up calling this method recursively.

However, I wonder if `super` might work...

Comment (bhearsum) [reply src/taskgraph/util/archive.py:46]: Then again, this is copied from upstream, and I don't think I want to diverge for that more than needed :stuck_out_tongue:

**Pull request merged**